### ISSUE TITLE
Disaster Recovery through Backup and Restore for 2A

### ISSUE THREAD

**Issue by DinaBelova** [k0rdent/kcm#590]
**Goals**

Provide a way backup and restore a 2A deployment in the event of a disaster that impact the mothership customer

**_Definitions_**:
- Backup
    - Ability to Backup all configuration objects created and managed by 2A into offsite location
- Restore
    - Ability to Create configuration objects from a specific Backup
- Disaster Recovery
    - Ability to restore a 2A system on another mgmt cluster with using Restore capability, plus ensuring that clusters are not recreated or lost
- Rollback
    - Restore after a specific event (like a failed upgrade). Can be a manual documentation that an 2A Operator could follow along.

**Major Deliverables**
Deliver a Backup System as part of 2A with possibility to handle a loss of the mgmt cluster

**Who it benefits**
- Customers
    - Recover the 2A system in the even of a disaster

**Acceptance criteria**
- Ability to make a full backup of 2A and its related configuration objects offboard (outside of the mgmt cluster to handle a loss of mgmt cluster)
- Backups should be able to be triggered manually by the Platform Lead.
- Backups should be able to be configured to happen automatically (by default every 6hours, but Platform Lead can change this in a Configuration ( Management  object?) 
- Ability to restore a 2A configuration object from a backup from a specific point in time without expecting 2A to reconcile or recreate the cluster
- The ability to restore a fully working 2A management cluster, including hosted control planes and continue to manage its deployed Managed Clusters
- Documented best practices what to take into consideration when changing the 2A management cluster (like changing source IP address of the mgmt cluster)
- The possibility to roll back all 2A created objects to a version before an upgrade
- Before an upgrade of 2A is executed all objects managed by 2A should be backed up
- Rollback can be initiated by the Platform Lead after a failed or successful upgrade through manual documented steps
- Test Upgrade and Rollback with every new 2A release with all our templates and check if a rollback causes problems. 

**Assumptions**
Managed Clusters will not fail in the event that the 2A system fails.

- We assume that all 2A component can handle a restored object
    - → need to test each of them
- This will span multiple sprints
- Possibly assign an engineer on this to dig deep into backups/restore
- We can assume that the new recovery cluster needs to be completely empty, so 2A should not handle any situation where a cert-manager is already installed in the new management cluster, etc. (We might want to add such verification in the future, but right now I think we can "ignore" this.)

**Limitations**
- Hosted control planes excluded from the first backup/restore version as they depend on k0smotron work

**Out of scope**
- Setup of a new k8s mgmt cluster
- Backup of hosted control planes (will be handled by k0smotron team and not the 2A team)

**User stories**
- As a Platform Lead, I want to backup my 2A configuration so that I can restore it in the event of a disaster.
- As a Platform Lead, I want to restore a fully working 2A Management cluster from backup after a disaster

**Comment by DinaBelova** [comment]
@zerospiel potential tasks in this epic (please check / create those, copying here from the list originally prepared by @Schnitzel )

- Research & Verification
    - Need to do research of different Backup Solutions
        - Velero
        - others?
- Test Backup & Restore with just CAPI and CAPx providers (outside of 2A/HMC)
    - If there are problems we would need to check with the CAPx providers to solve them and possibly block the whole backup&restore epic
- Test Backup & Restore with sveltos (potential problem with agent to management connectivity in case of a changed management cluster)
- Start implementation of Backup&Restore within 2A
    - Possibly need to wait on majority of all the 2A APIs
